rudderlabs/rudder-sdk-kotlin
# feat: add network context plugins

## Description


- In this PR, we added the `Network` context plugins which add the following fields:
    - `bluetooth`: true/false. This field is skipped if the sample Android app does not have the relevant permission.
    - `carrier`: A string value denoting the carrier name. If flight mode is activated then it is set to `NA`.
    - `cellular`: true/false. If the internet connection is toggled on then this will be set to true, or else false. If the device is connected to the internet through the Wifi and a cellular connection is enabled, then also this flag will be set to true.
    - `wifi`: true/false.

**NOTE**: We have not used any deprecated method to obtain network states.

## Type of change


## Implementation Details


### NetworkInfoPlugin

- It constructs the network context payload.
- If `ACCESS_NETWORK_STATE` permission is added in the sample Android app then context is constructed else skipped.
- If `BLUETOOTH` permission is not added then this value is not added in the context. Refer to our event spec: https://www.rudderstack.com/docs/event-spec/standard-events/common-fields/
> the Bluetooth status is collected in the network.bluetooth field only if the Bluetooth permission is present in the app.

### NetworkUtils

- It is a utility class that provides information about the network state. 
- This class uses both a `NetworkCallbackUtils` and a `DefaultNetworkUtils` to get information about the network.
- The `NetworkCallbackUtils` is used to get information about the `cellular` and `wifi` connection state.
- The `DefaultNetworkUtils` is used to get information about the `carrier` and `bluetooth` connection state and also as a fallback for the `cellular` and `wifi` connection state.

### DefaultNetworkUtils

- A utility class that provides information about the `carrier`, `cellular`, `wifi` and `bluetooth` connection state.
- It acts as a fallback for the `cellular` and `wifi` connection state.
- It contains the logic that we have in our v1 SDK. So this has already been tested.

### NetworkCallbackUtils

- A utility class that uses `ConnectivityManager.registerNetworkCallback` to register for `network` and `cellular` callbacks.
- This `registerNetworkCallback` can throw a `Runtime` exception if more than 100 callbacks are registered per app.
- We are only registering two callbacks, so users are free to register 98 callbacks.
- As a backup option, we are using the old v1 logic as a fallback in case `registerNetworkCallback` throws an exception.
- I've also noticed during testing that `cellular` is only set when `wifi` is disabled. So, we are forced to use the fallback way when this is set to `false`.

## Checklist


## How to test?


1. Toggle on/off: Bluetooth, WiFi, and Cellular connection.
2. Use a real device for checking the cellular info.
3. Turn on flight mode (carrier should be set to `NA`).
    a. **NOTE**: `Cellular` will be set to true/false depending upon its state when flight mode is enabled. If it was enabled and then flight mode is activated then it'll be set to `true` else vice-versa.
5. Edge case: Turn on both WiFi and Cellular (both fields should be set to `true`).
6. Remove the `Bluetooth` permission from the sample Android app.
7. Make events when the device is not connected to the internet.

When everything is turned on:
```
"network": {
    "bluetooth": true,
    "carrier": "T-Mobile",
    "cellular": true,
    "wifi": true
}
```

When flight mode is enabled:
```
"network": {
    "bluetooth": false,
    "carrier": "NA",
    "cellular": false,
    "wifi": false
}
```

When bluettoth permission is not set in the sample Android app:
```
"network": {
    "carrier": "T-Mobile",
    "cellular": true,
    "wifi": true
}
```

Full payload:
```
{
    "anonymousId": "2b1e22f0-74cd-4722-8b33-1a65aadb39ef",
    "channel": "mobile",
    "context": {
        "app": {
            "build": "100",
            "name": "Rudder-Android-Libs",
            "namespace": "com.rudderstack.android.sampleapp",
            "version": "0.1.0"
        },
        "library": {
            "name": "com.rudderstack.android.sdk",
            "version": "1.0.0"
        },
        "network": {
            "bluetooth": true,
            "carrier": "T-Mobile",
            "cellular": true,
            "wifi": true
        }
    },
    "event": "Application Opened",
    "integrations": {
        "All": true
    },
    "messageId": "82536b06-0604-4e3f-8211-1418ff4e1428",
    "originalTimestamp": "2024-10-02T21:01:50.598Z",
    "properties": {
        "from_background": false,
        "version": "0.1.0"
    },
    "receivedAt": "2024-10-02T21:01:56.385Z",
    "request_ip": "49.37.35.25",
    "rudderId": "4d04be17-869d-4e31-8b6a-3f5ba5698321",
    "sentAt": "2024-10-02T21:01:55.423Z",
    "type": "track"
}
```

## Breaking Changes


## Maintainers Checklist


## Screenshots (if applicable)

<img width="259" alt="image" src="https://github.com/user-attachments/assets/d744512f-3adc-4ca7-8253-b322d0813e98">

## Additional Context


References:

- https://developer.android.com/develop/connectivity/network-ops/reading-network-state.
- https://developer.android.com/reference/android/net/ConnectivityManager
